src/uu/wc/src/wc.rs

```diff
@@ -157,151 +248,165 @@ fn is_word_separator(byte: u8) -> bool {
-                    // -L is a GNU 'wc' extension so same behavior on LF
-                    longest_line_length = current_char_count - (ends_lf as usize);
+                Err(..) => {
+                    word_count += raw_line.split(|&x| is_word_separator(x)).count();
+                    current_char_count = raw_line.iter().filter(|c| c.is_ascii()).count()
                 }
             }
-
-            raw_line.truncate(0);
+            char_count += current_char_count;
+            if current_char_count > longest_line_length {
+                // -L is a GNU 'wc' extension so same behavior on LF
+                longest_line_length = current_char_count - (ends_lf as usize);
+            }
         }
 
-        results.push(Result {
-            title: path.clone(),
-            bytes: byte_count,
-            chars: char_count,
-            lines: line_count,
-            words: word_count,
-            max_line_length: longest_line_length,
-        });
+        raw_line.truncate(0);
+    }
 
-        total_line_count += line_count;
-        total_word_count += word_count;
-        total_char_count += char_count;
-        total_byte_count += byte_count;
+    Ok(WordCount {
+        bytes: byte_count,
+        chars: char_count,
+        lines: line_count,
+        words: word_count,
+        max_line_length: longest_line_length,
+    })
+}
 
-        if longest_line_length > total_longest_line_length {
-            total_longest_line_length = longest_line_length;
+fn word_count_from_path(path: &String, settings: &Settings) -> WcResult<WordCount> {
+    if path == "-" {
+        let stdin = io::stdin();
+        let stdin_lock = stdin.lock();
+        return Ok(word_count_from_reader(stdin_lock, settings, path)?);
+    } else {
+        let path_obj = Path::new(path);
+        if path_obj.is_dir() {
+            return Err(WcError::IsDirectory(path.clone()));
+        } else {
+            let file = File::open(path)?;
+            return Ok(word_count_from_reader(file, settings, path)?);
         }
+    }
+}
+
+fn wc(files: Vec<String>, settings: &Settings) -> Result<(), u32> {
+    let mut total_word_count = WordCount::default();
+    let mut results = vec![];
+    let mut max_width: usize = 0;
+    let mut error_count = 0;
+
+    let num_files = files.len();
 
-        // used for formatting
-        max_width = total_byte_count.to_string().len() + 1;
+    for path in &files {
+        let word_count = word_count_from_path(&path, settings).unwrap_or_else(|err| {
+            show_error!("{}", err);
+            error_count += 1;
+            WordCount::default()
+        });
+        max_width = max(max_width, word_count.bytes.to_string().len() + 1);
+        total_word_count += word_count;
+        results.push(word_count.with_title(path));
     }
 
     for result in &results {
-        print_stats(settings, &result, max_width);
+        print_stats(settings, &result, max_width).unwrap();
```

Comment:
Not sure if this is the best way to deal with this (since it’s not quite the same as the other errors), but I think rather than unwrapping, this should just add to `error_count` or something like that.  Maybe even immediately return with `Err(error_count + 1)`?

Reply:
`println!` unwraps internally, right? So isn't this essentially the same as before? If I remember correctly, errors writing to stdout are so rare that it's a failure condition that is almost never handled in C code (Have you ever checked the return code of `printf`?)

Reply:
Yeah, I don’t think the previous behavior was correct.

You can trigger an error by writing to a broken pipe.  It’s not super common, but it’s not like it can never occur.  I think most C programs just pretend the write doesn’t fail, so we could also just do that.  Unwrapping triggers a panic though, which IMO is overkill here (it also only works correctly because we catch panics and ignore pipe errors, which avoids the nasty panic error message).

Reply:
How do you like https://github.com/uutils/coreutils/pull/1851/commits/5c5e8a57e108619674a7d29b775e3344189afed2?

Reply:
Acutally, I should increment the error count as well...

Reply:
At the moment, it’ll fail if `stderr` also has a broken pipe, but I think this is probably good given that it’s not likely both `stdout` and `stderr` will have broken pipes.  I’ve been wondering if we should change the `show_*` macros to silently fail or simply abort rather than panic (this would reduce the size of the binaries and, I imagine, aborting would more closely mimic GNU coreutils).